### ISSUE TITLE
Best way to add general docs and help menu?

### ISSUE THREAD

**Issue by HagaiHargil** [napari/magicgui#107]
Hi :)

My main use case for magicgui is to replace command line applications with friendlier, GUI-based tools. This is very helpful for my non-coder colleagues, and magicgui has been invaluable on that front. One thing that I "miss" from CLI tools is documentation for the available parameters and more general "help documentation" which explains the use case that this current app is here to solve. In CLI tools these bits of information can usually be invoked with a `--help` flag which is obviously missing here.

I was wondering whether you think it would be possible to add text box for each parameter of the GUI which appears following a hover "event". This text box may pull its information from the docstring of the main annotated function (moar magik!). If you think it's doable I might try to dip my toes into this magical swamp.

In addition, a more traditional "Help" context menu or button can be optionally activated by adding a `"help_button: True"` entry to the main magicgui dictionary. This button can again get its information from the main body of the function's docstring.

How does it sound to you? I believe some docstring magic is already happening in order to decipher each type's designated widget.

**Comment by tlambert03** [comment]
> How does it sound to you?

sounds great!  it's definitely inline with the goal of making it easy for developers to provide _usable_ GUIs to people (and if it's not clear what all the widgets do, then it's not very useable! :joy:)

> I believe some docstring magic is already happening in order to decipher each type's designated widget.

we actually don't really use the docstrings much for type or widget inference (that all comes from either the type annotation, the default value, or the parameters to `magicgui()`)... however, you might be interested in the tooltips PR that recently merged! https://github.com/napari/magicgui/pull/100  (not yet released)

That should hopefully satisfy _part_ of the need here... but as you'll notice in that PR, the tooltips are restricted to the first sentence of the parameter docstring.  So a full-blown help would also be useful (maybe if `help_buton` is true, a little ℹ️ could appear somewhere in the widget that pops up the full docstring?)

**Comment by tlambert03** [comment]
I haven't added tooltips to the online docs yet... so until then, the best examples for tuning its behavior are probably [these tests](https://github.com/napari/magicgui/blob/45e7a15ac85f5e9a12249575c8efff80e0172d34/tests/test_magicgui.py#L483-L589)

**Comment by HagaiHargil** [comment]
Awesome, I missed that issue and the PR completely! I knew that "hover text box" wasn't the right phrase but I couldn't find a better one, thanks for these references.

I wonder why do you parse only the first sentence (up to the first comma or period)? Is there some technical limitation or do you think that it's simply better this way?

I believe that an :information_source: emoji is great, I'll try taking a deeper look.

**Comment by tlambert03** [comment]
> I wonder why do you parse only the first sentence (up to the first comma or period)? Is there some technical limitation or do you think that it's simply better this way?

It was a somewhat arbitrary decision, and with probably even a little nudging I could be convinced otherwise :).  No technical limitations.  Was just trying to keep the UI relatively clean... But maybe it was too strong of a choice.  A developer can always override the tooltips, as shown in those examples, so maybe we should just use the whole parameter docstring and leave it up to the dev to shorten manually if desired?

**Comment by HagaiHargil** [comment]
Honestly I can't think of a reason to cut off the tooltip this way a priori :X I'm not sure that a large tooltip text box will affect the cleanliness of the UI too much, because usability is more important than 'cleanliness' IMO, and we're talking about a disappearing tooltip - one that the developer can override quite easily.

**Comment by tlambert03** [comment]
sounds good, I'm convinced! :)  Any interest in submitting a PR to remove these two lines of splits:
https://github.com/napari/magicgui/blob/45e7a15ac85f5e9a12249575c8efff80e0172d34/magicgui/function_gui.py#L30-L31

and then update the tests correspondingly?
(if not, I can take care of it too)

**Comment by HagaiHargil** [comment]
Sure, I'll do it. Thanks!

**Comment by HagaiHargil** [comment]
Regarding that :information_source: widget - I'll need a `QIcon` widget, at least for the Qt backend. Do you think I should add it in `_concrete.py` and then import it to and from `widgets.py` or is there some other preferred way?

**Comment by tlambert03** [comment]
yeah, that's a good question.  It's not immediately clear how this is going to work across different backends.  Maybe one idea, since the implementation here likely to be rather backend-dependent, is that we add just one `_mgui_provide_help_icon` method to `magicgui.widgets._protocols.ContainerProtocol`... then you can do whatever you want inside the `magicgui.backends._qtpy.widgets.Contaniner._mgui_provide_help_icon` implementation to create a `QIcon` and add it to `self._layout`.  Then in `magicgui.widgets._bases.ContainerWidget`, we would have a method:

```python
def set_help_text(self, text: str = None):
    # if text is `None`, backends should remove the icon.
    self._widget._mgui_provide_help_icon(text)
```

thoughts?

**Comment by HagaiHargil** [comment]
I'm still exploring the parameter space here, as I'm not sure that this type of button is considered best practice. Perhaps a "Help -> About" button in the top app menu, that will only be shown if the help button is set to True is a better idea?

**Comment by tlambert03** [comment]
> Perhaps a "Help -> About" button in the top app menu, that will only be shown if the help button is set to True is a better idea?

Sure I like that, however... there is no help menu yet.  For that, we'd need to make widgets a [`QMainWindow`](https://doc.qt.io/qt-5/qmainwindow.html) and hookup some of the necessary functionality.

Most of what exists here was written with the idea of small composable widgets to be plugged into other things, and less as the whole application itself.  Perhaps we need another `main_window=True` flag or something for widgets that want to be the entire application unto themselves?  That would require more tooling to hook it up properly, but would likely be more extensible in the long run.

**Comment by tlambert03** [comment]
> I'm not sure that this type of button is considered best practice.

(I agree with you by the way, this would probably be the kind of thing that was annoying over time)

**Comment by HagaiHargil** [comment]
I see... I keep forgetting that magicgui's prime use case is napari :laughing: 

I guess it seems like a pretty big chunk of work needed for a very small benefit, and I'm not sure I can do it myself. Perhaps the tooltips are good enough after all.

**Comment by tlambert03** [comment]
> I keep forgetting that magicgui's prime use case is napari 

it's really not! and I don't want it to be seen that way.  I'd just say that was the original motivating use case.  Your use case is equally exciting to me.
I'm more than happy to look at implementing this.  I'll let you know if I hit on something.

**Comment by HagaiHargil** [comment]
Thanks for your effort on this (and other) front. One more thing, is this issue somehow related to the title bar of the window? Currently it simply says "magicgui" but I guess it should have the name of the decorated function. 

BTW, I really love magicgui and I believe you should publicize it more. The API seems to be converging to a very user-friendly place and the feature set is quite extensive already.

**Comment by tlambert03** [comment]
> Currently it simply says "magicgui" but I guess it should have the name of the decorated function.

I haven't really paid attention to that yet  (title bar test doesn't show on mac... are you on something else?).  But I agree, it should be the decorated function (or explicitly set with "name=" or something).  I'll look at that with the main window changes

> BTW, I really love magicgui and I believe you should publicize it more. The API seems to be converging to a very user-friendly place and the feature set is quite extensive already.

I really appreciate that.  I'm also happy with the recent directions and excited for the future of the project.  we do need to publicize more, you're right.  but i always feel like "wait! one more thing, and then I will!"

**Comment by HagaiHargil** [comment]
I'm on Linux, I see a "magicgui" title in Ubuntu and Arch under both X and Wayland.

I get what you're saying - your name is splashed all over the place and you want it to be great :)

**Comment by tlambert03** [comment]
closing after #110.  Will open a new issue for the window title

**Comment by uschmidt83** [comment]
It seems that the discussion here is (vaguely) related to what I'd like to use for my napari plugin, namely a little help icon next to an input field. I was reminded of this by @liaprins-czi's [presentation](https://docs.google.com/presentation/d/10pxnwvBBb1rYV-LEWgmn5b2n9CK93-Qc_U_CEKUD5jI/edit?usp=sharing) yesterday, and I also like the way it was implemented in [her example](https://docs.google.com/presentation/d/10pxnwvBBb1rYV-LEWgmn5b2n9CK93-Qc_U_CEKUD5jI/edit#slide=id.g118d275534b_0_195).

Is this already possible with magicgui? If not, any chance such (or similar) functionality might be added?
(If this is the wrong place to this discuss this, feel free to move it elsewhere.)

Best,
Uwe

**Comment by liaprins-czi** [comment]
If helpful, here's a link to the napari GH issue where I worked on the design for that little "?" info icon:
[Add pop-out button for grid mode and store defaults in settings #3009](https://github.com/napari/napari/issues/3009)
